how-to/workspace-platform-starter/client/src/framework/workspace/notification-client.ts

```diff
@@ -0,0 +1,211 @@
+	type UpdatableNotificationOptions,
+	type NotificationCreatedEvent,
+	create as notificationCreate,
+	update as notificationUpdate,
+	getAll as notificationGetAll,
+	clear as notificationClear,
+	show as notificationShow,
+	hide as notificationHide,
+	addEventListener as notificationAddEventListener,
+	removeEventListener as notificationRemoveEventListener
+} from "@openfin/workspace/notifications";
+import { isEmpty, randomUUID } from "workspace-platform-starter/utils";
+import type {
+	NotificationClient as NotificationClientInterface,
+	NotificationClientOptions
+} from "../shapes/notification-shapes";
+/**
+ *
+ */
+export class NotificationClient implements NotificationClientInterface {
+	private readonly _options: NotificationClientOptions;
+
+	private readonly _idPrefix: string;
+
+	private readonly _platformId: string;
+
+	private _listenerRegister: { id: string; listener: never; wrappedListener: never }[] = [];
+
+	/**
+	 * Notification options.
+	 * @param options Notification options.
+	 * @param platformId The id of the current platform this client will belong to.
+	 */
+	constructor(options: NotificationClientOptions, platformId: string) {
+		this._options = options;
+		this._idPrefix = options.idPrefix ?? `${options.id}-`;
+		this._platformId = platformId;
+	}
+
+	/**
+	 * Create a notification.
+	 * @param options the notification to create.
+	 * @param creationOptions creationOptions for the notification.
+	 * @returns created notification.
+	 */
+	public async create<T extends NotificationOptions>(
+		options: T,
+		creationOptions?: NotificationCreationOptions
+	): Promise<Notification<T>> {
+		if (!this.hasId(options.id)) {
+			options.id = this.createId(options.id);
+		}
+		if (this._options.enforceIcon && !isEmpty(this._options.icon)) {
+			options.icon = this._options.icon;
+		}
+
+		if (this._options.includeInPlatform) {
+			options.platform = this._platformId;
+		} else if (options.platform === this._platformId) {
+			options.platform = undefined;
+		}
+
+		return notificationCreate(options, creationOptions);
+	}
+
+	/**
+	 * Allows the updating of an existing notification.
+	 * @param options the notification to update.
+	 * @returns the updated notification.
+	 */
+	public async update<T extends UpdatableNotificationOptions>(options: T): Promise<Notification> {
+		if (!this.hasId(options.id)) {
+			options.id = this.createId(options.id);
+		}
+
+		return notificationUpdate(options);
+	}
+
+	/**
+	 * Get all notifications.
+	 * @returns all the notifications.
+	 */
+	public async getAll(): Promise<Notification[]> {
+		const notifications = await notificationGetAll();
+		return notifications.filter((entry) => entry.id.startsWith(this._idPrefix));
+	}
+
+	/**
+	 * Clears a specific notification.
+	 * @param id the id of the notification to clear.
+	 * @returns whether the notification was cleared.
+	 */
+	public async clear(id: string): Promise<boolean> {
+		const notificationToClear = this.createId(id);
```

Comment:
Why does this call createId ?

Reply:
This is on the off chance they had a local id, called create and then passed that same id again without checking the returned notification. This just adds the prefix if it is missing.

Reply:
But if they pass an empty string so that it allocates a random id, won't calling notificationClear will be trying to remove a non existent notification

Reply:
Notification clear needs an id. If they pass empty then we generate one and that is what will be in the returned notification. If they pass that id to clear then the create just re-uses it if it already has the required prefix